controllers/storagecluster/cephcluster.go

```diff
@@ -531,11 +531,18 @@ func newStorageClassDeviceSets(sc *ocsv1.StorageCluster, serverVersion *version.
 					if noPreparePlacement {
 						in := getPlacement(sc, "osd-prepare-tsc")
 						(&in).DeepCopyInto(&preparePlacement)
+					}
 
-						if len(topologyKeyValues) >= replica {
-							// If topologyKey is not host, append additional topology spread constarint to the
-							// default preparePlacement. This serves even distribution at the host level
-							// within a failure domain (zone/rack).
+					if len(topologyKeyValues) >= replica {
+						if portable && !strings.Contains(topologyKey, "zone") {
+							placement.TopologySpreadConstraints = append(placement.TopologySpreadConstraints, placement.TopologySpreadConstraints[0])
+							placement.TopologySpreadConstraints[0].TopologyKey = topologyKey
+							placement.TopologySpreadConstraints[0].WhenUnsatisfiable = "DoNotSchedule"
```

Comment:
Why not always set this? Without this, the only guard we would have against moving the OSDs across failure domains is the PV constraints. In some environments, even if we're running in zones, it's possible that the PVs might be schedulable across zones, although maybe not a likely scenario.

Reply:
My hope would be that the PVs would always come with constraints on where they can be used. Is this not the case?

Reply:
Verified most of the cloud providers that their PVs hold the constraints in case of zone failure domain. Also, in case of zones the OSD pod placement is dependent on prepare job placement as the PV created from it holds the actual constraint. Keeping this into consideration, we apply soft constraints for zones.

Reply:
@jarrpa For example, OCS puts the artificial constraints on racks. LSO isn't aware of the racks, so OCS needs to put constraints on the pods so they don't move across those boundaries we've chosen. Rook doesn't touch the PVs, it only consumes them through the PVCs. 
@kesavanvt My question is really why we shouldn't apply this constraint to zones. The constraint wouldn't really be needed for zones since the PVs have the constraints, but it shouldn't hurt either. Consistent code is easier to maintain than special cases. I just don't see the need for the special case here.

Reply:
IIRC we proceeded on `ScheduleAnyway`(soft constraint) keeping both "rook-ceph-osd-prepare", "rook-ceph-osd" as label selector for prepare pod placement, as topology spread only takes into consideration on the running pods and not on the completed jobs. It may be possible that it breaks the even spread at some point which is not likely to happen. Even if it does, the subsequent osd expansion would try to balance the even spread. Keeping this into account, having hard constraint for the OSD pod would be a problem for zones as it depends on pv constraint which is again dependent on prepare pod placement.

Reply:
This latest reasoning works for me. 👍